tests/numpy/test_gen_data.py

```diff
@@ -273,3 +273,25 @@ def test(arr):
 def test_inferring_from_time_dtypes_gives_same_dtype(data, dtype):
     ex = data.draw(nps.from_dtype(dtype))
     assert dtype == ex.dtype
+
+
+@given(st.data(), nps.byte_string_dtypes() | nps.unicode_string_dtypes())
+def test_inferred_string_strategies_roundtrip(data, dtype):
+    # Check that we never generate too-long or nul-terminated strings, which
+    # cannot be read back out of an array.
+    arr = np.zeros(shape=1, dtype=dtype)
+    ex = data.draw(nps.from_dtype(arr.dtype))
+    arr[0] = ex
+    assert arr[0] == ex
+
+
+@given(st.data(), nps.scalar_dtypes())
+def test_all_inferred_scalar_strategies_roundtrip(data, dtype):
+    # We only check scalars here, because record/compound/nested dtypes always
+    # give an array of np.void objects.  We're interested in whether scalar
+    # values are safe, not known type coercion.
+    arr = np.zeros(shape=1, dtype=dtype)
+    ex = data.draw(nps.from_dtype(arr.dtype))
+    assume(ex == ex)  # NaN is uninteresting for this test
```

Comment:
Maybe worth noting in the comment that we're using this test because we might have non-numeric types where `np.isnan` would error.